### ISSUE TITLE
twisted.protocols.test.test_tls fails when SSL is not available

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3852]
|<img alt="azverkan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| azverkan reported|
|-|-|
|Trac ID|trac#3852|
|Type|defect|
|Created|2009-05-23 02:36:24Z|
|Branch|https://github.com/twisted/twisted/tree/skip-if-ssl-missing-3852|

This test failure was encountered with win32 Python 2.6

```
[ERROR]: twisted.protocols.test.test_tls

Traceback (most recent call last):
  File "c:\work\twisted\twisted\trial\runner.py", line 640, in loadByNames
    things.append(self.findByName(name))
  File "c:\work\twisted\twisted\trial\runner.py", line 450, in findByName
    return reflect.namedAny(name)
  File "c:\work\twisted\twisted\python\reflect.py", line 462, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "c:\work\twisted\twisted\protocols\test\test_tls.py", line 24, in <module>
    from twisted.internet.ssl import ClientContextFactory, PrivateCertificate
  File "c:\work\twisted\twisted\internet\ssl.py", line 46, in <module>
    from OpenSSL import SSL
exceptions.ImportError: No module named OpenSSL
```

Attachments:

* [twisted_missing_ssl_test_skip.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/60f/60fbc1486c6584d80e5be72a9153098cd7f78ca8/984b81ccc4a9bbd8033161b0bf08894f49fa3226.patch) (1275 bytes) - added by azverkan on 2009-05-23 02:36:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3852 3852
type__defect defect
reporter__azverkan azverkan
priority__highest highest
milestone__ 
branch__branches_skip_if_ssl_missing_3852 branches/skip-if-ssl-missing-3852
branch_author__exarkun__azverkan exarkun, azverkan
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1243046184000000 1243046184000000
changetime__1246627674000000 1246627674000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph removed owner|
|-|-|

Thanks!

However, as described on TwistedDevelopment, please attach the 'review' keyword to the tickets which have a patch which you would like to have applied.  (And leave them unassigned, unless there's some an unusual reason that you need a specific person to review your contribution.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Fixing description markup.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)|@zooko set owner to @zooko<br>@zooko set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_4"></a>|@zooko commented|
|-|-|

Maciej and I reviewed the patch and it looks good.  However, the py2.5-without-modules builder is currently correctly skipping this test, instead of getting an import error.  The original bug report was apparently being run on Windows, and there is not currently a Windows builder which is testing without modules.

So the buildbot is not currently reproducing this problem on this platform.

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_5"></a>|@fijal commented|
|-|-|

I guess right now the status is more "not reproducible". What's exactly are the conditions
under which it can be reproduced?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I see the problem reproduced here: http://buildbot.twistedmatrix.com/builders/py2.5-without-modules/builds/169/steps/trial/logs/problems

Where did you look to see it skipped?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [26974]) Apply twisted_missing_ssl_test_skip.patch

refs [#3852](https://github.com/twisted/twisted/issues/3852)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [26973]) Branching to 'skip-if-ssl-missing-3852'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [26975]) Instead of handling import errors from twisted.internet.ssl as well, do them in the code block which only runs if OpenSSL is available

refs [#3852](https://github.com/twisted/twisted/issues/3852)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Applied the patch to a branch, with a slight change.  [Results on without-modules builder](http://buildbot.twistedmatrix.com/builders/py2.5-without-modules/builds/170).

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_11"></a>|@fijal commented|
|-|-|

Looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to @exarkun|
|-|-|

Merge, please?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [27084]) Merge skip-if-ssl-missing-3852

Author: azverkan, exarkun
Reviewer: fijal
Fixes: [#3852](https://github.com/twisted/twisted/issues/3852)

Be careful when importing ssl-related modules in the tests for
twisted.protocols.tls so that they skip cleanly when OpenSSL is
not available.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
